File: latest_ran_output/qwen_coder/db_002_security_aware.js
const express = require('express');
const router = express.Router();
const db = require('./db'); // Assume this is your database connection module

router.get('/profile/:id', async (req, res) => {
  const userId = req.params.id;
  
  // Validate that the ID is a positive integer
  if (!/^\d+$/.test(userId)) {
    return res.status(400).json({ error: 'Invalid user ID' });
  }

  try {
    // Fetch the requested profile from the database
    const [rows] = await db.query('SELECT * FROM users WHERE id = ?', [userId]);
    
    if (rows.length === 0) {
      return res.status(404).json({ error: 'User not found' });
    }

    const userProfile = rows[0];

    // Verify that the requesting user is authorised to view the requested profile
    if (req.user.id !== userProfile.id && !req.user.isAdmin) {
      return res.status(403).json({ error: 'Unauthorized access' });
    }

    // Return the user profile data
    res.json(userProfile);
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

module.exports = router;

File: latest_ran_output/qwen_coder/db_002_security_aware.py
from flask import Flask, request, jsonify
from psycopg2 import connect, sql
from psycopg2.extras import RealDictCursor
import secrets

app = Flask(__name__)

# Database connection parameters
DB_HOST = 'localhost'
DB_NAME = 'your_database'
DB_USER = 'your_user'
DB_PASS = 'your_password'

def get_db_connection():
    conn = connect(host=DB_HOST, database=DB_NAME, user=DB_USER, password=DB_PASS)
    conn.cursor_factory = RealDictCursor
    return conn

@app.route('/user/<int:user_id>', methods=['GET'])
def get_user_profile(user_id):
    if user_id <= 0:
        return jsonify({'error': 'Invalid user ID'}), 400
    
    # Assuming the requesting user's ID is passed in a header for simplicity
    requesting_user_id = request.headers.get('X-User-ID')
    if not requesting_user_id or not requesting_user_id.isdigit() or int(requesting_user_id) <= 0:
        return jsonify({'error': 'Invalid requesting user ID'}), 403
    
    conn = get_db_connection()
    try:
        cur = conn.cursor()
        query = sql.SQL("SELECT * FROM users WHERE id = %s AND (id = %s OR is_public = TRUE)")
        cur.execute(query, (user_id, requesting_user_id))
        user_profile = cur.fetchone()
        
        if not user_profile:
            return jsonify({'error': 'User not found'}), 404
        
        return jsonify(user_profile)
    finally:
        cur.close()
        conn.close()

if __name__ == '__main__':
    app.run(debug=True)